lib/common/pool.c

```diff
@@ -58,25 +60,25 @@ struct POOL_ctx_s {
 };
 
 /* POOL_thread() :
-   Work thread for the thread pool.
-   Waits for jobs and executes them.
-   @returns : NULL on failure else non-null.
-*/
+ * Work thread for the thread pool.
+ * Waits for jobs and executes them.
+ * @returns : NULL on failure else non-null.
+ */
 static void* POOL_thread(void* opaque) {
     POOL_ctx* const ctx = (POOL_ctx*)opaque;
     if (!ctx) { return NULL; }
     for (;;) {
         /* Lock the mutex and wait for a non-empty queue or until shutdown */
         ZSTD_pthread_mutex_lock(&ctx->queueMutex);
 
-        while (ctx->queueEmpty && !ctx->shutdown) {
+        while ( ctx->queueEmpty
+            || (ctx->numThreadsBusy >= ctx->threadLimit) ) {
+             if (ctx->shutdown) {
```

Comment:
What if `ctx->shutdown && !ctx->queueEmpty && ctx->numThreadsBusy >= ctx->threadLimit`? Might it drop work in this case?

Reply:
Good point !
Yes, it might.

I'm surprised it passes the new test, because this case is supposed to happen.

Reply:
I think that condition is only possible when the pools is downsized.